stdlib/crypto/src/main/ballerina/crypto/crypto.bal

```diff
@@ -0,0 +1,187 @@
+#
+# + path - Path to the key store file
+# + password - Key store password
+public type KeyStore record {
+    string path = "";
+    string password = "";
+    !...;
+};
+
+# Record for providing trust store related configurations.
+#
+# + path - Path to the key store file
+# + password - Key store password
+public type TrustStore record {
+    string path = "";
+    string password = "";
+    !...;
+};
+
+# Private key used in cryptographic operations.
+#
+# + algorithm - Key algorithm
+public type PrivateKey record {
+    KeyAlgorithm algorithm;
+    !...;
+};
+
+# Public key used in cryptographic operations.
+#
+# + algorithm - Key algorithm
+public type PublicKey record {
+    KeyAlgorithm algorithm;
+    !...;
+};
+
+# Error relevant to crypto operations.
+#
+# + message - Error message
+public type CryptoError record {
+    string message = "";
+    !...;
+};
+
+# Returns the MD5 hash of the given data.
+#
+# + input - Value to be hashed
+# + return - Hashed output
+public extern function hashMd5(byte[] input) returns byte[];
+
+# Returns the SHA-1 hash of the given data.
+#
+# + input - Value to be hashed
+# + return - Hashed output
+public extern function hashSha1(byte[] input) returns byte[];
+
+# Returns the SHA-256 hash of the given data.
+#
+# + input - Value to be hashed
+# + return - Hashed output
+public extern function hashSha256(byte[] input) returns byte[];
+
+# Returns the SHA-384 hash of the given data.
+#
+# + input - Value to be hashed
+# + return - Hashed output
+public extern function hashSha384(byte[] input) returns byte[];
+
+# Returns the SHA-512 hash of the given data.
+#
+# + input - Value to be hashed
+# + return - Hashed output
+public extern function hashSha512(byte[] input) returns byte[];
+
+# Returns the HMAC using MD-5 hash function of the given data.
+#
+# + input - Value to be hashed
+# + key - Key used for HMAC generation
+# + return - HMAC output
+public extern function hmacMd5(byte[] input, byte[] key) returns byte[];
+
+# Returns the HMAC using SHA-1 hash function of the given data.
+#
+# + input - Value to be hashed
+# + key - Key used for HMAC generation
+# + return - HMAC output
+public extern function hmacSha1(byte[] input, byte[] key) returns byte[];
+
+# Returns the HMAC using SHA-256 hash function of the given data.
+#
+# + input - Value to be hashed
+# + key - Key used for HMAC generation
+# + return - HMAC output
+public extern function hmacSha256(byte[] input, byte[] key) returns byte[];
+
+# Returns the HMAC using SHA-384 hash function of the given data.
+#
+# + input - Value to be hashed
+# + key - Key used for HMAC generation
+# + return - HMAC output
+public extern function hmacSha384(byte[] input, byte[] key) returns byte[];
+
+# Returns the HMAC using SHA-512 hash function of the given data.
+#
+# + input - Value to be hashed
+# + key - Key used for HMAC generation
+# + return - HMAC output
+public extern function hmacSha512(byte[] input, byte[] key) returns byte[];
```

Comment:
Why not returning error as well from these methods?

Reply:
The HMAC implementation throws "NoSuchAlgorithmException" if the algorithm is not supported. We know that we support all the algorithms we have exposed (since we have limited number of algorithms). Then it throws "InvalidKeyException" if the key is empty. That is anyway not allowed when passing data to native function. Therefore, there is no logical way we will end up in a error condition with HMAC functions.

Reply:
Same applies for Hash native functions. There is no way we will encounter "NoSuchAlgorithmException", since we know that the algorithm we have selected are supported in Ballerina environments.